processor/src/trace/virtual_bus/circuit/prover.rs

```diff
@@ -0,0 +1,471 @@
+    let mut merged_mls = vec![poly_a, poly_b, poly_c, poly_d, poly_x];
+
+    // run the first sum-check protocol
+    let ((round_claim, before_merge_proof), r_sum_check) = sum_check_prover_plain_partial(
+        claimed_evaluation,
+        num_rounds_merge,
+        &mut merged_mls,
+        transcript,
+    );
+
+    // parse the output of the first sum-check protocol
+    let RoundClaim {
+        eval_point: rand_merge,
+        claim,
+    } = round_claim;
+
+    // create the composed multi-linear for the second sum-check protocol using the randomness
+    // sampled during the first one
+    let gkr_composition =
+        gkr_merge_composition_from_composition_polys(&composition_polys, r_sum_check, rand_merge);
+
+    // include the partially evaluated at the first sum-check randomness EQ multi-linear
+    mls.push(merged_mls[4].clone());
+
+    // run the second sum-check protocol
+    let main_prover = SumCheckProver::new(gkr_composition, SimpleGkrFinalClaimBuilder(PhantomData));
+    let after_merge_proof = main_prover.prove(claim, mls, transcript).unwrap();
+
+    FinalLayerProof {
+        before_merge_proof,
+        after_merge_proof,
+    }
+}
+
+/// Proves all GKR layers except for input layer.
+fn prove_before_final_circuit_layers<
+    E: FieldElement<BaseField = Felt> + 'static,
+    C: RandomCoin<Hasher = H, BaseField = Felt>,
+    H: ElementHasher<BaseField = Felt>,
+>(
+    circuit: &mut FractionalSumCircuit<E>,
+    transcript: &mut C,
+) -> (BeforeFinalLayerProof<E>, GkrClaim<E>) {
+    // absorb the circuit output layer. This corresponds to sending the four values of the output
+    // layer to the verifier. The verifier then replies with a challenge `r` in order to evaluate
+    // `p` and `q` at `r` as multi-linears.
+    let num_layers = circuit.p_0_vec.len();
+    let data = vec![
+        circuit.p_0_vec[num_layers - 1][0],
+        circuit.p_1_vec[num_layers - 1][0],
+        circuit.q_0_vec[num_layers - 1][0],
+        circuit.q_1_vec[num_layers - 1][0],
+    ];
+    // generate the challenge `r`
+    transcript.reseed(H::hash_elements(&data));
+
+    // generate the challenge and reduce [p0, p1, q0, q1] to [pr, qr]
+    let r = transcript.draw().unwrap();
+    let mut claim = circuit.evaluate_output_layer(r);
+
+    let mut proof_layers: Vec<SumCheckProof<E>> = Vec::new();
+    let mut rand = vec![r];
+    for layer_id in (1..num_layers - 1).rev() {
+        // construct the Lagrange kernel evaluated at the previous GKR round randomness
+        let poly_x = EqFunction::ml_at(rand.clone());
+
+        // construct the vector of multi-linear polynomials
+        // TODO: avoid unnecessary allocation
+        let poly_a = circuit.p_0_vec[layer_id].to_owned();
+        let poly_b = circuit.p_1_vec[layer_id].to_owned();
+        let poly_c = circuit.q_0_vec[layer_id].to_owned();
+        let poly_d = circuit.q_1_vec[layer_id].to_owned();
+        let mut mls = vec![poly_a, poly_b, poly_c, poly_d, poly_x];
```

Comment:
Should be changed

Reply:
Right, putting the `EqFunction` as the last `ml` is very brittle, and makes the code harder to follow. Ultimately, we need the `EqFunction` evaluation in the `GkrComposition::evaluate()`; couldn't we just evaluate it there directly?

And then `SimpleGkrFinalClaimBuilder` wouldn't need to remove the last opening.